coderd/externalauth/externalauth.go

```diff
@@ -542,7 +562,43 @@ func applyDefaultsToConfig(config *codersdk.ExternalAuthConfig) {
 	}
 }
 
-var defaults = map[codersdk.EnhancedExternalAuthProvider]codersdk.ExternalAuthConfig{
+func bitbucketServerDefaults(config *codersdk.ExternalAuthConfig) codersdk.ExternalAuthConfig {
+	defaults := codersdk.ExternalAuthConfig{
+		DisplayName: "Bitbucket Server",
+		Scopes:      []string{"PUBLIC_REPOS", "REPO_READ", "REPO_WRITE"},
+		DisplayIcon: "/icon/bitbucket.svg",
+	}
+	// Bitbucket servers will have some base url, e.g. https://bitbucket.coder.com.
+	// We will grab this from the Auth URL. This choice is a bit arbitrary,
+	// but we need to require at least 1 field to be populated.
+	if config.AuthURL == "" {
+		// No auth url, means we cannot guess the urls.
+		return defaults
+	}
+
+	auth, err := url.Parse(config.AuthURL)
+	if err != nil {
+		// We need a valid URL to continue with.
+		return defaults
+	}
+
+	// Populate Regex, ValidateURL, and TokenURL.
+	// Default regex should be anything using the same host as the auth url.
+	defaults.Regex = fmt.Sprintf(`^(https?://)?%s(/.*)?$`, strings.ReplaceAll(auth.Host, ".", `\.`))
```

Comment:
Shouldn't we be doing these for all of them? Seems like making it just for BitBucket server is weird.

Reply:
perhaps the self-hosted ones.

Reply:
We do.

Azure
https://github.com/coder/coder/blob/84509a19de051f0a9fd43c7bd770cfd373c96ba6/coderd/externalauth/externalauth.go#L607-L607

Bitbucket Cloud
https://github.com/coder/coder/blob/84509a19de051f0a9fd43c7bd770cfd373c96ba6/coderd/externalauth/externalauth.go#L616-L616


Gitlab
https://github.com/coder/coder/blob/84509a19de051f0a9fd43c7bd770cfd373c96ba6/coderd/externalauth/externalauth.go#L625-L625

Github
https://github.com/coder/coder/blob/84509a19de051f0a9fd43c7bd770cfd373c96ba6/coderd/externalauth/externalauth.go#L634-L634

Bitbucket Server is always hosted on a different domain for each customer. So it's a dynamic regex string based on the `AUTH_URL`. The other providers are all cloud based with static urls.

Reply:
@spikecurtis yes, the self hosted ones need this dynamic approach. We need to handle them case by case. Maybe there will be generalizations, but for now I only implemented self hosted bitbucket server.

Gitlab for example can definitely have some broken defaults right now.

Reply:
@Emyrk automatically doing the Regex should be a global change in my mind; it's weird to do it for a single provider.

Reply:
The regex is automatic for all providers. As in, the default values are correct for every type now

Reply:
Talked to @kylecarbs offline a bit.

Yes this problem exists with Gitlab and Github Enterprise. We should probably have some solution for these.

I do not want to break backwards compatibility, and I do not have the test environments to confirm those (eg, are the urls consistent? bitbucket cloud and server are different).

So for this PR, I'd rather not try and solve github enterprise and gitlab self hosted.